### ISSUE TITLE
Give developers control over "overlay" browser

### ISSUE THREAD

**Issue by cabanier** [immersive-web/webxr#1365]
Meta Quest recently introduced support to show the 2d browser while staying in VR/AR.
This mode can only entered/left by the user using system gestures.

We've gotten multiple requests since launching this features to allow for developer control over this. Basically, developers want to trigger overlay by calling an API so they can show a configuration or shopping screen. Then they want to present the user with a button to resume the WebXR session.

/facetoface give developers control over "overlay" browser

**Comment by KooIaIa** [comment]
On other platforms like PC it is common to be able to open a web browser overlay inside of an XR application without freezing it. Instead of having this new developer control halt the WebXR session - could you also have an option to let the overlay browser be accessible inside the WebXR session? I think developers would love being able to access the overlay browser from inside a WebXR session and control it.

**Comment by cabanier** [comment]
> Instead of having this new developer control halt the WebXR session - could you also have an option to let the overlay browser be accessible inside the WebXR session? I think developers would love being able to access the overlay browser from inside a WebXR session and control it.

That's what this proposal is

**Comment by Utopiah** [comment]
Would it be a single overlay or multiple overlays? Would the overlay (or overlays) be fixed or positional?

I'd prefer multiple positional overlays.

**Comment by cabanier** [comment]
It's just one and it will always be centered. Basically what you have today except you can trigger it as an author

**Comment by hybridherbst** [comment]
It could also be interesting to have drag-and-drop functionality between that browser window and the XR session (and back?).

**Comment by josephrocca** [comment]
There's the potential for something really powerful here - more than just a userland API for the current system function.

It'd be very useful to be able to open multiple browser windows *while still being active in the current session*.

I think the browser window(s) would have to locally (i.e. in spatially-aware manner) "capture" user interaction from the parent WebXR session - kinda like an iframe does for mouse/keyboard/etc. events in HTML.

So e.g. could seamlessly tap pause button on YouTube window while remaining in game/social world. And if it's just another browser tab then you could stream it to your friends in social VR using media capture API. This sort of paradigm feels like it leans into the web's strengths.

Perhaps too architecturally ambitious for now, but I hope the spec for this can leave some possibilities open for future proposals. It's a *very* exciting direction!

**Comment by KooIaIa** [comment]
> It's just one and it will always be centered. Basically what you have today except you can trigger it as an author

Today, a VR browser user is able to size, position, and place a browser html window wherever they want (almost). Imagine opening a website in AR passthrough that opens WebXR. Could the page keep its original user set position and then just be visible in WebXR? Picture clicking the Enter XR button and the webpage never disappears and cleanly transitions into WebXR as the default behavior.

>There's the potential for something really powerful here - more than just a userland API for the current system function.

The powerful feature I am most excited for is detatched 3D CSS. Once a HTML page can be loaded in VR like this, the next step could be adding VR support to the web's existing 3D functionality to make it immersive XR 3D. Then standard HTML and CSS wil work in VR and allow placing as many 3D overlays as a page could want and work in 2D too.

**Comment by coderofsalvation** [comment]
Sounds promising, and I realize there are always security-implications.
However, from a user and developer point of view I'd like the author to be able to enable basic features:

* enable URL-bar
* enable bookmarks
* enable back/forward buttons
* enable text-select/copy/paste
* unrestricted mode: just overlay the browserwindows from homescreen (worstcase enabled via browser flag)

> Reason: the web requires knowledge-navigation: even triggering an payment-link will in most cases require the user to look up info elsewhere (having to exit immersive mode to open a new tab would defeat the purpose of the overlay imho).

Unrestricted mode would resonate with doing serious work in (Web)XR.

**Comment by josephrocca** [comment]
Potentially relevant viral (1.7k likes) [tweet](https://twitter.com/DX_Nacca/status/1770775580754981125) in the VRChat community from yesterday:

> DXOverlay, Show your friends when you have the Overlay open. Integration with XSOverlay directly into VRChat via OSC. With easy customization of colors and screens.

![image](https://github.com/immersive-web/webxr/assets/1167575/62d336f5-a46e-45af-9c72-549c34925d51)

I'm guessing this feature could be in MVP/v1.0 of this spec - just requires exposing position, orientation, and size of each overlay browser/panel, which is kind of a bare minimum anyway if this proposal goes beyond `toggleOverlay()` at all.

**Comment by josephrocca** [comment]
A security consideration: If the user's hands are near the keyboard that's associated with an overlay, hand/finger tracking information should not be exposed to WebXR, since this could leak passwords and other info. Might also need to consider lowering eye and arm tracking resolution - in the extreme case (probably not necessary), the user's body could completely "freeze" from WebXR's perspective while they're interacting with an overlay.

As mentioned earlier, this is akin to the browser iframe (or OS window) capturing mouse/keyboard events - usually based on the spatial position of the input device (though there is also the concept of frame "focus" in browser/OS land - not sure if that's relevant here - maybe gaze factors into that).

**Comment by cabanier** [comment]
I'd like to bring this feature back. Several developers have asked us if they can bring up the browser without quitting the experience. I vaguely remember that we discussed this since 2024 but I wasn't able to find it in the minutes.

My proposal is to have a "toggleOverlay" (or "showOverlay") as proposed [here](https://github.com/immersive-web/webxr/issues/1365#issuecomment-2015197007) that will show the browser on top of a visible-blurred WebXR session. UA that don't support this may exit the session.

/agenda discuss "toggleOverlay"

**Comment by coderofsalvation** [comment]
Having visible-blurred as a conditional feature would be great,
Usecases like 'non-distraction' or secure-mode would benefit from it (`.html` with javascript e.g.).
But if it's just a readonly URL (rendering a `.txt` or `.md` file etc) then non-blurred WebXR environment would be great (pull up documentation etc).
I'd suggest white-listing certain readonly filetypes (`.txt`, `.md`,..) to bypass visible-blurring of WebXR surroundings. 
A webbrowser is not only for interactive HTML-content (!)

**Comment by AdaRoseCannon** [comment]
I don't think we would need visible-blurred on VisionOS unless hand-tracking is enabled because transient-pointer inputs only exist during an interaction and we can suppress the ones that would hit the window. So it's good to make a requirement but only when there are inputs that could reveal information.

**Comment by KooIaIa** [comment]
>non-blurred WebXR environment would be great (pull up documentation etc).

Blurring could be on while inputs are toggled / suppressed and then blurring could be turned off when input is off so it could be used for information / music / watching videos like this.

**Comment by cabanier** [comment]
Comments from the meeting:
- Apple asked to make visible-blurred an option if the UA uses tracked-pointer and not hands or other constant inputs.
- toggleOverlay sounds like a reasonable name for the API

**Comment by cabanier** [comment]
Proposal: #1439
